src/main/java/com/commercetools/sync/commons/utils/StreamUtils.java

```diff
@@ -32,38 +27,29 @@
     }
 
     /**
-     * Resolves all not empty optionals of a given {@link Stream} and returns them as a {@link List}.
+     * Similar to Java 9
+     * <a href="https://docs.oracle.com/javase/9/docs/api/java/util/Optional.html#stream--">Optional#stream()</a>:
+     * converts Optional to a stream of one element if not empty. Otherwise - return an empty stream.
      *
-     * @param elements the stream of elements.
-     * @return a {@link List} of resolved optionals.
+     * @param optional Optional to convert
+     * @param <T>      type of optional value
+     * @return Stream of one element from the supplied {@code optional} if not empty, otherwise - an empty stream.
      */
     @Nonnull
-    public static <T extends List<?>> T asList(@Nonnull final Stream<? extends Optional> elements) {
-        return filterStream(elements, Collectors.toList());
+    public static <T> Stream<T> filterEmptyOptionals(final @Nonnull Optional<T> optional) {
+        return optional.map(Stream::of).orElseGet(Stream::empty);
     }
 
     /**
-     * Resolves all not empty optionals of a given {@link Stream} and returns them as a {@link Set}.
+     * Similar to {@link #filterEmptyOptionals(Optional)}, but returns a lambda with the same behavior.
      *
-     * @param elements the stream of elements.
-     * @return a {@link Set} of resolved optionals.,
-     */
-    @Nonnull
-    public static <T extends Set<?>> T asSet(@Nonnull final Stream<? extends Optional> elements) {
-        return filterStream(elements, Collectors.toSet());
-    }
-
-    /**
-     * Resolves all not empty optionals of a given {@link Stream} and returns them as a Collection<T>.
-     *
-     * @param elements the stream of elements.
-     * @param  {@link Collector}, that accumulates input elements into a mutable result container
-     * @return a {@link Collection<T>} of resolved optionals.
+     * @param <T> type of optional value
+     * @return function which returns a stream of one element from the supplied {@code optional} if not empty,
+     *     otherwise - an empty stream.
      */
     @Nonnull
-    private static <T extends Collection> T filterStream(@Nonnull final Stream<? extends Optional> elements, @Nonnull
-        Collector<Object, ?, ?> collector) {
-        return (T) elements.filter(Optional::isPresent).map(Optional::get).collect(collector);
+    public static <T> Function<Optional<? extends T>, Stream<? extends T>> getFilterEmptyOptionals() {
```

Comment:
I think it's ok to explicitly write the type of the generic with method reference. So I vote not to keep this util

Reply:
Not sure what do you mean, what's wrong with generics in this method?
the only one thing i noticed - redundant `? extends`: b56194e3

Reply:
No, I didn't mean the generics in this method signature. I meant the need for this method at all. As you mentioned already:

> for better readability static import of `getFilterEmptyOptionals()` could be used instead of `StreamUtils::filterEmptyOptionals` reference. Compare this and this with other util calls in the same test class which use `StreamUtils::filterEmptyOptionals`

I prefered to have the consistent style of using method references with stream api and lambda's. i.e. using `StreamUtils::filterEmptyOptionals` instead of `getFilterEmptyOptionals()` even if we have to explicitly mention the generic types of the stream elements, like here: [`Stream.<Optional<String>>of`](https://github.com/commercetools/commercetools-sync-java/pull/275/files#diff-75070d8e3eb2924f08d9b46edc1ab484R92) I think it's totally ok, readability-wise.

Reply:
e93ecdc0